만국박람회 [STEP2] Hisop, Charles

Hi-sop opened the pull request:
@1Consumption  
STEP2 PR입니다.

## 고민했던 점

### NSMutableAttributedString 사용
    
첫 화면의 "방문객", "개최지", "개최 기간"의 폰트 크기가 다른것을 확인해 수정해보고자 했습니다.
- Label의 일부 스트링이 다른 속성을 가지고 있도록 설정하고 싶었습니다.
- 3개의 Label에 적용시켜주어야 했으므로 UILabel에 extension 메서드를 추가해 사용하도록 했습니다.
    
    
```swift
extension UILabel {
    func changeFontSize(targetString: String) {
        let attributedString = NSMutableAttributedString(string: self.text ?? "")
        attributedString.addAttribute(.font, value: UIFont.systemFont(ofSize: 20), range: ((self.text ?? "") as NSString).range(of: targetString))
        
        self.attributedText = attributedString
    }
}
```
```swift
visitorsLabel.changeFontSize(targetString: "방문객")
```

### setContentOffset / scrollRectToVisible 스크롤 이동에 쓰이는 메서드 고민
예시 화면에서 backBarButtonItem을 이용하여 첫 화면으로 돌아올 때 스크롤의 위치가 최상단으로 변경되는 것을 확인해 적용시켜주고자 했습니다.
- setContentOffset
    스크롤 뷰를 전달받은 CGPoint, 한 점으로 이동하는 메서드라 이해했습니다.
- scrollRectToVisible
    스크롤 뷰를 전달받은 CGRect, 직사각형이 표시되도록 이동하는 메서드라 이해했습니다.

특정 포인트를 지정해주기보단 titleLabel을 보여줄것이다 라는 표현이 더 적절하다 생각하여 scrollRectToVisible을 선택했습니다.
```swift
scrollView.scrollRectToVisible(titleLabel.frame, animated: true)
```

### defaultContentConfiguration
[모던 셀 구성 - WWDC](https://developer.apple.com/videos/play/wwdc2020/10027/)를 접하게되어 셀 구성에 해당 기술을 써보고자 했습니다.
```swift
var content = cell.defaultContentConfiguration()
        
content.image = UIImage(named: culturalAssets[indexPath.row].imageName) 
content.text = culturalAssets[indexPath.row].name
content.secondaryText = culturalAssets[indexPath.row].shortDescription
        
cell.contentConfiguration = content
```
위 방식으로 셀의 이미지, 텍스트를 설정할 수 있었습니다.

- 셀 내부의 이미지 정렬
처음에는 커스텀 셀을 만들어 세부적인 이미지의 크기와 레이아웃을 정해주는 쪽으로 수정할지 고민했습니다.
```swift
content.imageProperties.maximumSize = CGSize(width: 60, height: 60)
content.imageProperties.reservedLayoutSize = CGSize(width: 65, height: 0)
```
maximumSize로 이미지의 크기를, reservedLayoutSize로 레이아웃을 정해주도록 설정했습니다.

### Label과 TextView 사용을 고민했습니다.
텍스트 뷰에 스크롤 기능이 내장되어 있는 것을 확인하고 채택해보았습니다. 텍스트 뷰 위에 이미지 뷰를 올려보니 텍스트 뷰 하위에 속하지 않아 텍스트만 스크롤 되고 이미지는 고정되어 있어 스크롤 뷰에 Label을 추가하는 방식으로 구현하였습니다.

### 네비게이션 hidden!
- 네비게이션 바를 숨기기 위한 방법은 무엇이 있는 지 알아보았습니다. 우선 `setNavigationBarHidden(_, animated:)` 메서드가 있었고, `isNavigationBarHidden`라는 속성이 있었습니다. 두 가지 방법을 사용해 보았고 두 방법에 대한 차이는 바가 사라지는 것을 애니메이션으로 보여줄 수 있는 지에 대한 여부라 판단되어 애니메이션 기능자체를 사용하지 않을 거라 후자를 선택하였습니다.

## 조언을 얻고 싶었던 점

### attributedString / NSMutableAttributedString의 차이
- 문자열의 특정 부분의 속성만 바꾸기 위해 알아보았던 방법들이 `AttributedString`과 `NSMutableAttributedString`이었습니다. 처음 구현은 `AttributedString`으로 시도하였고, Label 텍스트로 넣어주는 부분에서 실패를 하여 검색을 해보았지만 SwiftUI로 구현된 방법밖에 제시되어 있지 않아 `NSMutableAttributedString`으로 구현하였습니다. 
- AttributedString 구현할 수 있는 방법이 있는 지에 대해 조언을 얻고 싶습니다.

### Decoding 시행하는 시점에 대한 고민
현재 CulturalAsset에 대한 Decoding은 CulturalAssetListViewController의 ViewDidLoad에 포함되어있어 화면이 전환될때마다 Decoding이 시행됩니다.
현재는 변경되지 않는 JSON파일을 관리하게 되는데 이 과정이 불필요해보여 수정을 고민하고 있습니다.

네트워크로 받는다고 가정할 때에는 어떤 식으로 Decoding을 처리하게 되는지 궁금합니다.
JSON이 변경되었을 때만 다시 디코딩을 하게 되나요?

### 스토리보드를 통한 화면전환 / 코드를 통한 화면전환 선택

현재 명확히 선택하는 기준을 찾지 못했습니다. 선택하는 기준이 있을까요?

Commit:
feat: 네비게이션 컨트롤러 생성 및 segue 연결

Charl-es committed:
feat: 뷰컨트롤러 위에 스크롤 뷰 생성

Commit:
feat: StackView및 Label, image 생성

Charl-es committed:
rename: 첫 화면을 나타내는 ViewController 네이밍 변경

Commit:
feat: ViewController에 Label, imageView 연결

Charl-es committed:
feat: 데이터를 디코딩하는 메서드 생성

Commit:
feat: configureUI메서드 추가

Charl-es committed:
refactor: 스크롤 뷰 위의 뷰 삭제

Commit:
feat: StackView 오토레이아웃 추가

Charl-es committed:
feat: 하단의 버튼과 이미지뷰 스택 뷰로 묶어서 추가

Commit:
feat: buttonImages 추가

Charl-es committed:
feat: TableViewController 생성과 객체연결

Commit:
feat: Button과 CulturalAssetListViewController 연결, Button 네이밍 수정

Charl-es committed:
feat: CulturalAssetListViewController 네비게이션 뒤로가기 바 버튼 재설정

Commit:
feat: CulturalAssetListViewController에 decodeDataAsset 메서드 추가

Charl-es committed:
feat: items JSON을 디코딩한 데이터를 생성한 셀에 할당

Commit:
feat: DetailViewController 추가

Charl-es committed:
feat: DetailVieewController에 텍스트 뷰 추가

Commit:
feat: didSelectRowAt를 통하여 Cell클릭시 화면 전환 구현

Charl-es committed:
feat: 텍스트 뷰에 이미지 뷰 추가

Commit:
feat: 화면전환시 Detail 뷰컨트롤러의 프로퍼티에 값 전달하는 코드 추가

Charl-es committed:
refactor: 텍스트 뷰 삭제, 스크롤 뷰와 스택 뷰 추가

Commit:
feat: 전달받은 데이터 화면에 출력하는 코드 추가

Commit:
refactor: 첫번째 화면 전환 방식 코드베이스로 변경

Charl-es committed:
feat: Expostion의 타이틀 개행을 위한 연산 프로퍼티 구현

Hi-sop committed:
feat: visitors의 값을 formatting하는 연산프로퍼티 구현

Charl-es committed:
refactor: Exposition의 titleWithNewLine 연산 프로퍼티 네이밍 변경

Hi-sop committed:
feat: location의 값을 formatting하는 연산프로퍼티 구현

Charl-es committed:
feat: Exposition에 formattingduration 연산프로퍼티 구현

Hi-sop committed:
feat: 화면 전환시 scrollView가 최상단을 바라볼 수 있도록 하는 코드 추가

Charl-es committed:
feat: UILabel 확장을 통해 일부 문자열의 폰트 크기를 변경하는 메서드 구현

Hi-sop committed:
feat: 첫 화면에서 네비게이션 컨트롤러가 보이지 않도록 수정

Charl-es committed:
refactor: 마지막 화면에서 이전 화면으로 돌아가는 바 버튼 아이템 타이틀 수정

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+
+import UIKit
+
+class CulturalAssetListViewController: UITableViewController {
```
`CulturalAssetListViewController`는 상속되는 클래스인가요? 아니라면 어떤 키워드로 `상속되지 않음`을 보여줄 수 있을까요?

Hi-sop replied:
상속가능성이 없는 class들에 final을 추가했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+import UIKit
+
+class CulturalAssetListViewController: UITableViewController {
+    var culturalAssets:[CulturalAsset] = []
```
외부에서 알아야하는 프로퍼티인가요?

Hi-sop replied:
class내부의 접근제어를 수정했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+import UIKit
+
+class CulturalAssetListViewController: UITableViewController {
+    var culturalAssets:[CulturalAsset] = []
```
```suggestion
    var culturalAssets: [CulturalAsset] = []
```

공백이 빠졌네요.

Hi-sop replied:
공백을 추가했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+        guard let detailViewController = storyboard?.instantiateViewController(identifier: String(describing: DetailViewController.self)) as? DetailViewController else {
+            return
+        }
+        navigationItem.backBarButtonItem = UIBarButtonItem(title: "한국의 출품작", style: .plain, target: self, action: nil)
```
이 코드는 `DetailViewController`에서 해야할 역할 같습니다.

Hi-sop replied:
뒤로가기 버튼 설정을 '현재 뷰컨트롤러'에서 진행하는게 적절하다고 생각하여 ```navigationController?.navigationBar.topItem?.title```을 수정하는 방식으로 변경하였습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+        
+        detailViewController.name = culturalAssets[indexPath.row].name
+        detailViewController.imageName = culturalAssets[indexPath.row].imageName
+        detailViewController.detailDescription = culturalAssets[indexPath.row].detailDescription
```
이런 정보를 프로퍼티로 하나하나 할당하지 말고, 하나의 펑션으로 받아서 처리하면 어떨까요?

Charl-es replied:
말씀해 주신 부분을 참고하여 두 객체 간의 의존성을 낮추기 위해서 `DetailViewController`의 프로퍼티에 값을 할당해 주는 메서드를 구현하였고 이를 `CulturalAssetListViewController`에서 호출하는 방식으로 수정하였습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+
+        guard let dataAsset = NSDataAsset(name: "items") else {
+            return
+        }
```
`NSDataAsset(name: "items")` 생성에 실패하면 `JSONDecoder`가 필요 없기 때문에, 코드의 순서 변경이 필요해보입니다.

Hi-sop replied:
생각하지 못하고 습관에 의해 작성한 부분인 것 같습니다. 순서 변경했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+        do {
+            culturalAssets = try decoder.decode([CulturalAsset].self, from: dataAsset.data)
+        } catch {
+            print(error.localizedDescription)
```
만약 에러가 발생한다면 유저가 알아차릴 수 있을까요?

Charl-es replied:
해당 부분은 에러가 발생했을 경우 얼럿을 띄워주고 이전 화면으로 돌아가도록 구현하였습니다.

1Consumption commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,30 @@
+    var imageName: String = ""
+    var detailDescription: String = ""
+    @IBOutlet weak var imageView: UIImageView!
+    @IBOutlet weak var explanation: UILabel!
```
외부에서 알 필요가 없으니 private으로 바꿔주세요. 또한 굳이 강제 언래핑을 하지 않아도 될 것 같습니다.
물론 자동 생성된 코드지만, 본인들이 커밋한 코드니까 자동생성된 코드에도 본인들의 기준이 들어가면 좋겠습니다.

Hi-sop replied:
뷰 컨트롤러의 ViewDidLoad가 호출되기 전에는 IBOutlet은 nil이 들어있고, 뷰가 생성되고 난 뒤로부터는 해당 뷰컨트롤러에 항상 값이 들어있을 것이기 때문에 암시적으로 언래핑한 옵셔널으로 선언되어 있는 것으로 알고있었습니다.

해당 코드 부분을 코멘트해주신 이유를 찾아보았습니다. 
```swift
//DetailViewController
func changeValue(test: String) {
        explanation.text = test
    }

//CulturalAssetListViewController
detailViewController.changeValue(test: "TEST")
```
암시적 언래핑 옵셔널로 선언되었을 때에는 위 상황과 같이 ViewDidLoad가 아직 호출되지 않아 IBOutlet의 값에 nil이 들어있는 상태에서 접근을 시도하면 런타임에러가 발생하는 것을 확인했습니다.

현재 저희 코드에서도 ViewDidLoad호출 전에 IBOutlet에 접근하는 부분이 없으므로 그대로 사용해주었습니다.

1Consumption commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,30 @@
+        imageView.image = UIImage(named: imageName)
+        
+        explanation.text = detailDescription
+    }
```
외부에서 알 필요가 없는 메소드 같습니다.

Hi-sop replied:
해당 메서드의 접근제어를 수정했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,30 @@
+    }
+}
+
+
```
불필요한 공백이네요.

Hi-sop replied:
수정완료했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/FirstViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        configureUI()
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        scrollView.scrollRectToVisible(titleLabel.frame, animated: true)
+        navigationController?.isNavigationBarHidden = true
+    }
```
override메소드인데, super를 호출하지 않아도 되나요?

Hi-sop replied:
놓친 부분인 것 같습니다. 호출하도록 변경했습니다.

override 메소드만 실행하게 되면 super에서 필수적으로 실행되야하는 코드들이 실행되지 않을 가능성이 있으므로 
super를 호출하는 것이 맞으며, 일부 예외가 되는 메서드도 있으므로 공식문서 기준으로 작성하면 된다고 이해했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/FirstViewController.swift:
```diff
@@ -0,0 +1,77 @@
+
+import UIKit
+
+class FirstViewController: UIViewController {
```
`FirstViewController`라는 이름은 여러가지 해석의 여지가 있을 것 같습니다. 좀 더 명확한 이름을 붙이는건 어떨까요?

Charl-es replied:
ExpositionIntroductionViewController로 수정해 보았습니다.

1Consumption commented on Expo1900/Expo1900/Controller/FirstViewController.swift:
```diff
@@ -0,0 +1,77 @@
+            return nil
+        }
+
+        return try? decoder.decode(Exposition.self, from: dataAsset.data)
```
이 구조의 코드가 위에서도 쓰인 것 같은데, 클래스나 구조체로 따로 분리하고, 여기서 사용하면 어떨까요?

Hi-sop replied:
AssetParser 구조체를 만들어 decodeDataAsset 메서드를 이동 및 수정시켜 주었습니다.

이 과정에서 기존에 시행하던 do-try-catch구문이 viewDidLoad로 이동했습니다. viewDidLoad에서 에러에 따른 동작을 시행하도록 하는 것은 적절했을까요? 구조체의 메서드로 / 뷰컨트롤러의 메서드로 분리해서 처리해주는것이 더 적절했을까요?

1Consumption replied:
에러 처리는 사용하는 제일 끝단에서 하는게 좋습니다. 그 이전까지는 에러를 래핑하거나, 넘길 뿐이죠.

1Consumption commented on Expo1900/Expo1900/Controller/FirstViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        
+        self.attributedText = attributedString
+    }
+}
```
여기서만 쓰이는 extension 같은데, 굳이 extension으로 만들지 않고 ViewController의 메소드로 만들어도 되지 않나요?

Charl-es replied:
해당 메서드는 ViewController 내부의 메서드로 구현하였습니다.

1Consumption commented on Expo1900/Expo1900/Controller/CulturalAssetListViewController.swift:
```diff
@@ -0,0 +1,66 @@
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let detailViewController = storyboard?.instantiateViewController(identifier: String(describing: DetailViewController.self)) as? DetailViewController else {
+            return
+        }
```
[`instantiateViewController(identifier:creator:)`](https://developer.apple.com/documentation/uikit/uistoryboard/3213989-instantiateviewcontroller)를 사용하면 init 시점에 `name`, `imageName`, `detailDescription`을 넣어 줄 수 있습니다.

Hi-sop replied:
```swift
func setUp(culturalAsset: CulturalAsset) {
        name = culturalAsset.name
        imageName = culturalAsset.imageName
        detailDescription = culturalAsset.detailDescription
    }
```
위 코멘트에서 setUp메서드를 만들고 호출하여 프로퍼티에 값을 주입해주는 방식으로 변경했습니다. 
이 방식과 init시점에 프로퍼티의 값들을 넣어주는것이 유사하다고 생각하여 수정하지 않기로 결정했습니다.

1Consumption replied:
저는 이부분에 대해서는 생각이 좀 다른데요, init 시점에 값을 넣어주지 않으면 아래와 같이 불필요한 값 할당이 필요해집니다. 사실은 바뀌지 않는 값들이지만 var로 선언해야 하죠,
``` swift
var name: String = ""
var imageName: String = ""
var detailDescription: String = ""
```

두가지 정도 선택지가 있을 것 같습니다.
1. init 시점에 값을 넣어준다.
2. 옵셔널로 변경해 불필요한 값을 넣어주지 않는다.

Hi-sop replied:
b2fcf02b8e5822112f05bcad358cb908c1be2166
제시해주신 선택지에서 1번으로 변경하였습니다.

관련하여 여러 테스트를 하던 중 개인적인 궁금증이 생겨 질문드립니다.
https://sonihemant111.medium.com/required-initializers-in-swift-4805de8ade9e
매개변수가 없다면 자동으로 super.init이 불리지만, 매개변수가 존재한다면 자동으로 불릴 수 없는 이유가 궁금합니다.

매개변수가 있는 init은 super init의 매개변수에 어떤 요소를 전달할지 몰라서 자동으로 호출할 수 없는걸까요?

1Consumption reviewed (changes requested):
고생하셨습니다. 코멘트를 참고해주세요.

# 질문
### attributedString / NSMutableAttributedString의 차이
- 문자열의 특정 부분의 속성만 바꾸기 위해 알아보았던 방법들이 `AttributedString`과 `NSMutableAttributedString`이었습니다. 처음 구현은 `AttributedString`으로 시도하였고, Label 텍스트로 넣어주는 부분에서 실패를 하여 검색을 해보았지만 SwiftUI로 구현된 방법밖에 제시되어 있지 않아 `NSMutableAttributedString`으로 구현하였습니다. 
- AttributedString 구현할 수 있는 방법이 있는 지에 대해 조언을 얻고 싶습니다.

# 답변
`NSAttributedString`의 [`init(_:)`](https://developer.apple.com/documentation/foundation/nsattributedstring/3787788-init) 메소드가 있는데, 이걸 말씀하시는 걸까요?

-------------------------------------------
# 질문
### Decoding 시행하는 시점에 대한 고민
현재 CulturalAsset에 대한 Decoding은 CulturalAssetListViewController의 ViewDidLoad에 포함되어있어 화면이 전환될때마다 Decoding이 시행됩니다.
현재는 변경되지 않는 JSON파일을 관리하게 되는데 이 과정이 불필요해보여 수정을 고민하고 있습니다.

네트워크로 받는다고 가정할 때에는 어떤 식으로 Decoding을 처리하게 되는지 궁금합니다.
JSON이 변경되었을 때만 다시 디코딩을 하게 되나요?

# 답변
> JSON이 변경되었을 때만 다시 디코딩을 하게 되나요?

JSON이 변경 되었다 라는게 정확히 어떤 것을 말씀하시는 걸까요? 지금은 asset에 고정된 데이터를 디코딩하니까 변경되지 않았다 라고 말씀하시는걸까요?

그렇다면 말씀하신대로 변경이 되지 않는 것이 맞는데 네트워킹을 대체한다는 가정을 하고 있으니, 네트워킹과 동일하게 생각해주시면 되겠습니다. 그렇다면 보통은 viewDidLoad에서 수행하거든요.
또는 [`ETag`](https://developer.mozilla.org/ko/docs/Web/HTTP/Headers/ETag)라는 것이 있긴 한데, 사용해본적은 없어서 이부분은 말씀을 드리기 애매하네요.

-------------------------------

# 질문
### 스토리보드를 통한 화면전환 / 코드를 통한 화면전환 선택

현재 명확히 선택하는 기준을 찾지 못했습니다. 선택하는 기준이 있을까요?

# 답변
기준은 정하기 나름입니다. 다만 일관성이 중요할 것 같습니다. 스토리보드를 사용하면 스토리보드만 사용하는식으로요.
또는 넘겨줘야 하는 데이터가 있다면 코드로 작성하는 것이 좋겠습니다.

Hi-sop commented:
AttributedString을 사용했을 때 String으로 변경해 UILabel.text에 넣으려고 했으나 변경이 불가능해
NSAttributedString을 사용하여 구현했습니다.

답변해주신 NSAttributedString의 init메서드로
AttributedString을 타입변경해 attributedText에 넣는 방법도 있다는 결론을 얻었습니다.

하지만 AttributedString만을 사용하기 위해 iOS버전을 15로 올리고 싶지 않아서 NSAttributedString을 사용하는것을 유지하는것으로 결정했습니다.

Hi-sop committed:
refactor: ViewController들에 final키워드를 추가

Charl-es committed:
refactor: 접근제어 수준 변경

Hi-sop committed:
refactor: buttonTapped접근제어 수준 변경, 불필요한 공백 제거

Charl-es committed:
style: CulturalAssetListViewController의 cultualAsset의 공백 추가

Hi-sop committed:
refactor: 뒤로가기 버튼의 구현 방식 변경

navigationItem.backBarButtonItem -> navigationController?.navigationBar.topItem?.title에 할당하는 방식으로

Charl-es committed:
refactor: DetailViewController의 프로퍼티 값을 할당해 주는 메서드 구현

Hi-sop committed:
refactor: decodeDataAsset에서 JSONDecoder 생성 시점 변경

적절한 dataAsset이 없을 경우 생성하지 않도록 위치를 변경

Charl-es committed:
JSON의 데이터를 디코딩 중 에러가 발생한다면 얼럿을 띄워주고 이전 화면으로 돌아가도록 구현

Hi-sop committed:
refactor: viewDidAppear의 super메서드를 호출하도록 변경

Charl-es committed:
refactor: 네이밍 수정

Hi-sop committed:
feat: decodeDataAsset메서드 AssetParser구조체로 이동 및 수정

Charl-es committed:
refactor: 확장에서의 메서드를 뷰컨트롤러의 메서드에 구현

1Consumption commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,36 @@
+        configureUI()
+    }
+
+    private func configureUI() {
+        navigationItem.title = name
+        imageView.image = UIImage(named: imageName)
+        
+        explanation.text = detailDescription
+    }
+    
+    func setUp(culturalAsset: CulturalAsset) {
+        name = culturalAsset.name
+        imageName = culturalAsset.imageName
+        detailDescription = culturalAsset.detailDescription
+    }
```
private 메소드와 internal 메소드의 순서가 따로 있을까요? 지금은 internal 메소드가 나오다가 private이 나오고, 다시 internal이 나와서 일관성이 없다고 느껴지네요.

Hi-sop replied:
https://github.com/yagom-academy/ios-exposition-universelle/commit/b2fcf02b8e5822112f05bcad358cb908c1be2166
구현을 변경하며 setUp메서드가 삭제되었습니다.

1Consumption commented on Expo1900/Expo1900/Controller/ExpositionIntroductionViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        } catch {
+            let alert = UIAlertController(title: error.localizedDescription, message: "앱이 종료됩니다.", preferredStyle: .alert)
+            let okAction = UIAlertAction(title: "예", style: .default) { action in
+                exit(0)
```
어떤 상황에도 앱을 종료해서는 안됩니다. 이부분은 UX의 문제라서 유저에게 안좋은 경험을 전달할 수 있겠습니다.

Charl-es replied:
https://github.com/yagom-academy/ios-exposition-universelle/pull/300/commits/d41c201c38224a4cae53af5935a3be30a3848681

이 부분에 대해서는 다시 고민을 해보았습니다.
만약 네트워크 환경이었으면 다시 데이터를 요청하는 함수를 구현하는 방식으로 시도를 했을 것 같으나,
현재 데이터에셋을 이용한 한정적인 상황이라 사용자에게 얼럿을 보여주는 것으로 충분하다 생각하였습니다.

1Consumption commented on Expo1900/Expo1900/Model/AssetParser.swift:
```diff
@@ -0,0 +1,35 @@
+    
+    var errorDescription: String? {
+        switch self {
+        case .dataAssetNameError:
+            return "dataAssetNameError"
+        }
```
여기는 사용자에게 전달될 에러 메세지를 적는게 좋을 것 같습니다. 유저들은 `dataAssetNameError`의 의미를 알지 못할 것 같아요.

Charl-es replied:
https://github.com/yagom-academy/ios-exposition-universelle/pull/300/commits/8f3a2e8c7feb18e6e7058be943c662dde386206b
유저가 알아보기 쉬운 문구로 수정하였습니다.

1Consumption commented on Expo1900/Expo1900/Model/AssetParser.swift:
```diff
@@ -0,0 +1,35 @@
+        
+        let decoder = JSONDecoder()
+        
+        do {
+            return try decoder.decode(T.self, from: dataAsset.data)
+        } catch {
+            throw error
+        }
```
어차피 decodeDataAsset(assetName:) 메소드가 throws 펑션인데, 굳이 do catch를 해준 이유가 있을까요?

Hi-sop replied:
596309513bab5ec55054eb83031000935501d591
Result타입을 반환하게 하며 같이 수정되었습니다.

1Consumption commented on Expo1900/Expo1900/Model/AssetParser.swift:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+struct AssetParser<T: Decodable> {
+    func decodeDataAsset(assetName: String) throws -> T {
+        guard let dataAsset = NSDataAsset(name: assetName) else {
+            throw dataAssetError.dataAssetNameError
+        }
+        
+        let decoder = JSONDecoder()
+        
+        do {
+            return try decoder.decode(T.self, from: dataAsset.data)
+        } catch {
+            throw error
+        }
+    }
+}
```
이 객체는 왜 struct로 만들었나요? class나 enum으로 만드는 방법도 있었을 것 같은데요.
또한 앱 전반적으로 사용되는 객체같은데, 사용할 때마다 인스턴스를 만들어서 사용하는 이유가 있을까요?

Hi-sop replied:
ad208156efaecf6ae6ce3fed9c57a7ca8f45782a
AssetParser를 enum으로 변경, decodeDataAsset에 static 키워드를 추가해주었습니다.

사용하는 메서드에 static키워드를 추가해 인스턴스화 시키지 않는다는 점은 같다고 생각합니다.
위 코드를 struct로 만드는 것과 enum으로 만드는 것은 어떤 차이가 있을까요?
저희는 private init을 추가해야한다는 차이만 느껴집니다.

1Consumption commented on Expo1900/Expo1900/Model/AssetParser.swift:
```diff
@@ -0,0 +1,35 @@
+}
+
+struct AssetParser<T: Decodable> {
+    func decodeDataAsset(assetName: String) throws -> T {
```
[`Result`](https://developer.apple.com/documentation/swift/result)타입을 사용해도 좋을 것 같습니다.

Hi-sop replied:
020653e48faaa8ccff68273ebb0c9fdb3b924881
decodeDataAsset이 Result타입을 반환하도록 변경했습니다.

1Consumption reviewed (changes requested):
코멘트를 참고해주세요.

Hi-sop committed:
refactor: DetailViewController의 init을 오버로딩하여 호출하도록 변경

instantiateViewController(identifier:creator:)사용하여 호출

Charl-es committed:
refactor: 첫 번째 화면에서 디코딩을 실패했을 때 앱이 종료되는 코드 삭제

Hi-sop committed:
refactor: DataAssetError및 AssetParser수정

AssetParser가 Result 리턴하도록 수정

Hi-sop committed:
refactor: decodeDataAsset을 호출하는 부분 수정

do-catch -> switch로 변경

Charl-es committed:
refactor: 상황에 맞는 에러 문구 출력 수정

Hi-sop committed:
refactor: AssetParser enum으로 수정 decodeDataAsset에 static키워드 추가

Hi-sop committed:
refactor: viewDidLoad의 switch 메서드로 분리
